src/magic.cpp

```diff
@@ -0,0 +1,804 @@
+#include "magic.h"
+
+#include "calendar.h"
+#include "color.h"
+#include "game.h"
+#include "generic_factory.h"
+#include "map.h"
+#include "mapdata.h"
+#include "messages.h"
+#include "monster.h"
+#include "mutation.h"
+#include "npc.h"
+#include "options.h"
+#include "player.h"
+#include "projectile.h"
+#include "translations.h"
+
+// LOADING
+// spell_type
+
+namespace
+{
+generic_factory<spell_type> spell_factory( "spell" );
+}
+
+template<>
+const spell_type &string_id<spell_type>::obj() const
+{
+    return spell_factory.obj( *this );
+}
+
+template<>
+bool string_id<spell_type>::is_valid() const
+{
+    return spell_factory.is_valid( *this );
+}
+
+void spell_type::load_spell( JsonObject &jo, const std::string &src )
+{
+    spell_factory.load( jo, src );
+}
+
+energy_type energy_source_from_string( std::string str )
+{
+    if( str == "MANA" ) {
+        return mana_energy;
+    } else if( str == "HP" ) {
+        return hp_energy;
+    } else if( str == "BIONIC" ) {
+        return bionic_energy;
+    } else if( str == "STAMINA" ) {
+        return stamina_energy;
+    } else {
+        return none_energy;
```

Comment:
I feel that invalid energy source and damage type strings should be detected and reported, rather than just defaulting to `none_energy` and `DT_TRUE`.  That will help people who mistype something in the json.